# Importing essential libraries

In [2]:
import pandas as pd
import numpy as np
import re
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix
from sklearn.utils.class_weight import compute_class_weight
import matplotlib.pyplot as plt
import seaborn as sns
import itertools
from collections import Counter
from extract_chr import extract_chr1
from sklearn.metrics import roc_curve
from prepare_fragments import get_fragments_df

In [3]:
import lightgbm as lgb

In [4]:
import sys
print(sys.executable)

d:\Projects\DNA Sequence Classifier\.venv\Scripts\python.exe


# Importing the dataset

In [5]:
extract_chr1(r"D:\Projects\DNA Sequence Classifier\Required_data\Homo_sapien.fna", "human_chr1.fna",[2])
extract_chr1(r"D:\Projects\DNA Sequence Classifier\Required_data\Canis_lupis_familiaris.fna", "dog_chr1_2.fna",[1,2])
#extract_chr1(r"D:\Projects\Required_data\Pan_troglodytes.fna", "chimp_chr1.fna")
human = get_fragments_df("human_chr1.fna", "Human",1000)
dog = get_fragments_df("dog_chr1_2.fna", "Dog",1000)
#chimp = get_fragments_df("chimp_chr1.fna", "Chimp",1000)

df = pd.concat([human,dog], ignore_index = True)

print(f"Total sequences: {len(df)}")
print(df['species'].value_counts())

Found primary chromosome 2 header : >NC_000002.12 Homo sapiens chromosome 2, GRCh38.p14 Primary Assembly
Chromosome 2 extraction complete.
Done.
Found primary chromosome 1 header : >NC_049222.1 Canis lupus familiaris isolate Mischka breed German Shepherd chromosome 1, alternate assembly UU_Cfam_GSD_1.0, whole genome shotgun sequence
Chromosome 1 extraction complete.
Found primary chromosome 2 header : >NC_049223.1 Canis lupus familiaris isolate Mischka breed German Shepherd chromosome 2, alternate assembly UU_Cfam_GSD_1.0, whole genome shotgun sequence
Chromosome 2 extraction complete.
Done.
Human:
Total length:  242508799
Total Fragments: 242508
Dog:
Total length:  208535887
Total Fragments: 208535
Total sequences: 451043
species
Human    242508
Dog      208535
Name: count, dtype: int64


# Visualizing the data before cleaning

In [6]:
print(df['sequence'])
print(f"\n{human.describe()}")
print(f"\n{df.describe()}")
print(f"\n{df.head}")
print(df.columns)
print(type(df))

0         NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...
1         NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...
2         NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...
3         NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...
4         NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...
                                ...                        
451038    TTCCCAGGGGAAGTAGGAAATTATACATATTCTATGGAAGCTCTTT...
451039    CATTAAGGAATCAATCCCATTTACAATTGCACCCAAAAGCATAAGA...
451040    CAACACTTGGAAACAAGTCCTGGTACTCATCACAGGTGCCCTCCAT...
451041    AAAGACATTTTTGCCTATCTTTCCTTCTTTTTATGGTTTCAAGTCT...
451042    TTCGCGGTGGCCGTTCGGGTTCGTGTTCGGGTTTCGTGTTCGGGGT...
Name: sequence, Length: 451043, dtype: object

                                                 sequence species
count                                              242508  242508
unique                                             240876       1
top     NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...   Human
freq                         

print(df)
for species in df['species'].unique():
    subset = df[df['species'] == species]
    plt.hist(subset['length'],alpha = 0.6, label = species)

plt.legend()
plt.xlabel("Sequence Length")
plt.ylabel("Frequency")
plt.title("DNA Length Distribution")
plt.show()

sns.boxplot(data = df, x = 'species', y = 'length')
plt.title("DNA Length Distribution by Species")

sns.violinplot(data=df, x = 'species', y = 'length')
plt.title("DNA Length Distribution by Species")
plt.show()

sns.violinplot(data=df, x = 'species', y = 'length')
plt.title("DNA Length Distribution by Species")
plt.show()

sns.barplot(data = df, x = 'species', y = 'length')
plt.title("Average DNA Length by Species")
plt.show()

# Cleaning the sequences

In [7]:
df = df[df['sequence'].str.count('N') < 50]
df['sequence'] = df['sequence'].str.upper()
print(f"\n{df.head}")
print(f"Clean sequences: {len(df)}")


<bound method NDFrame.head of                                                  sequence species
10      CGTATCCCACACACCACACCCACACACCACACCCACACACACCCAC...   Human
11      CGCGCAGGCGCAGAGAGGCGCACCGCGCCCGCGCAGGCGCAGAGAG...   Human
12      CTGAAACGGAGCTATTAGTGGGGAGAGCTGATGTCCCAGTTCTTGT...   Human
13      GCTATCAACTCACACAAAGACATGAGTGAATCTTGCATGCACATTG...   Human
14      CGTGGTGTTATCTCGCCTCACTGCAACCTCTGCCTTCTGGGTTCAC...   Human
...                                                   ...     ...
451038  TTCCCAGGGGAAGTAGGAAATTATACATATTCTATGGAAGCTCTTT...     Dog
451039  CATTAAGGAATCAATCCCATTTACAATTGCACCCAAAAGCATAAGA...     Dog
451040  CAACACTTGGAAACAAGTCCTGGTACTCATCACAGGTGCCCTCCAT...     Dog
451041  AAAGACATTTTTGCCTATCTTTCCTTCTTTTTATGGTTTCAAGTCT...     Dog
451042  TTCGCGGTGGCCGTTCGGGTTCGTGTTCGGGTTTCGTGTTCGGGGT...     Dog

[449347 rows x 2 columns]>
Clean sequences: 449347


# Visualizing the data after cleaning

In [8]:
print(f"Total sequences: {len(df)}")
print(df['species'].value_counts())

Total sequences: 449347
species
Human    240845
Dog      208502
Name: count, dtype: int64


# K-mer feature extraction

Function Definations

In [9]:
def build_kmer_vocab(k):
    alphabet = ['A','T','G','C']
    return [''.join(p) for p in itertools.product(alphabet, repeat = k)]

def kmer_vector(seq, k, kmer_index):
    counts = Counter(seq[i:i+k] for i in range(len(seq) - k + 1))
    total = sum(counts.values()) or 1
    vec = np.zeros(len(kmer_index), dtype = float)
    
    valid_total = 0
    for kmer, c in counts.items():
        if kmer in kmer_index:
            vec[kmer_index[kmer]] = c
            valid_total += c

    if valid_total > 0:
            vec = vec / valid_total

    n_ratio = seq.count('N') / len(seq)

    vec = np.append(vec, n_ratio)
    
    return vec

Calling the Methods

def call_function(k):
    vocab = build_kmer_vocab(k)
    kmer_index = {kmer: i for i,kmer in enumerate(vocab)}
    print(f"k = {k} creates {len(vocab)} features")

    #Converting all sequences to feature
    X = np.vstack(df['sequence'].apply(lambda s: kmer_vector(s, k, kmer_index)))
    y = pd.Categorical(df['species']).codes

    X_train, X_test, y_train, y_test = train_test_split(X,y,test_size = 0.2, random_state = 42, stratify = y)

    rf = RandomForestClassifier(n_estimators = 200, max_depth = None, min_samples_split = 5, n_jobs = -1, random_state = 42)
    rf.fit(X_train, y_train)

    y_pred = rf.predict(X_test)
    accuracy = accuracy_score(y_test, y_pred)
    print(f"\nAccuracy: {accuracy:.3f}")

    species_names = ['Chimpanzee','Dog','Human']
    print("\nClassification Report:")
    print(classification_report(y_test, y_pred, target_names = species_names))

for i in range(1,10):
    call_function(i)

In [10]:
k = 5
vocab = build_kmer_vocab(k)
kmer_index = {kmer: i for i,kmer in enumerate(vocab)}
print(f"k = {k} creates {len(vocab)} + 1(n_ratio) features")

#Converting all sequences to feature
X = np.vstack(df['sequence'].apply(lambda s: kmer_vector(s, k, kmer_index)))
y = pd.Categorical(df['species']).codes

k = 5 creates 1024 + 1(n_ratio) features


# Train/test split

In [11]:
X_train, X_test, y_train, y_test = train_test_split(X,y,test_size = 0.2, random_state = 42, stratify = y)

In [12]:
classes = np.unique(y_train)
weights = compute_class_weight(
    class_weight='balanced',
    classes=classes,
    y=y_train
)
class_weight_dict = dict(zip(classes, weights))
sample_weights = np.array([class_weight_dict[y] for y in y_train])

# Train LightGBM

In [13]:
train_data = lgb.Dataset(X_train, label = y_train, weight= sample_weights)
valid_data = lgb.Dataset(X_test, label = y_test)

In [53]:
parameters = {
    'objective' : 'binary',
    'metric' : 'auc',
    'num_leaves' : 80,
    'boosting' : 'gbdt',
    'learning_rate' : 0.05,

    'feature_fraction' : 0.5,
    'min_child_samples' : 20,
    'random_state' : 42,
    'verbosity': -1
}

In [54]:
model_lgbm = lgb.train(parameters,
                       train_data,
                       valid_sets=[train_data,valid_data],
                       valid_names=['train', 'valid'],
                       num_boost_round= 5000,
                       callbacks=[
                           lgb.early_stopping(15),
                           lgb.log_evaluation(50)
                           ]
)

Training until validation scores don't improve for 15 rounds
[50]	train's auc: 0.841975	valid's auc: 0.835791
[100]	train's auc: 0.862651	valid's auc: 0.851522
[150]	train's auc: 0.876486	valid's auc: 0.860674
[200]	train's auc: 0.886959	valid's auc: 0.866619
[250]	train's auc: 0.895301	valid's auc: 0.8708
[300]	train's auc: 0.902529	valid's auc: 0.873764
[350]	train's auc: 0.908825	valid's auc: 0.876157
[400]	train's auc: 0.914547	valid's auc: 0.878146
[450]	train's auc: 0.919698	valid's auc: 0.879684
[500]	train's auc: 0.924376	valid's auc: 0.880811
[550]	train's auc: 0.928781	valid's auc: 0.881935
[600]	train's auc: 0.93297	valid's auc: 0.882985
[650]	train's auc: 0.936796	valid's auc: 0.883692
[700]	train's auc: 0.940406	valid's auc: 0.884229
[750]	train's auc: 0.943804	valid's auc: 0.884665
[800]	train's auc: 0.947048	valid's auc: 0.88506
[850]	train's auc: 0.950056	valid's auc: 0.885353
[900]	train's auc: 0.953036	valid's auc: 0.885544
[950]	train's auc: 0.95579	valid's auc: 0.88

# Train Random Forest

rf = RandomForestClassifier(n_estimators = 150, max_depth = None, min_samples_split = 5, n_jobs = -1, random_state = 42, class_weight = "balanced")
rf.fit(X_train, y_train)

# Evaluating

from sklearn.metrics import confusion_matrix

cm = confusion_matrix(y_test, y_pred)
print(cm)

from sklearn.metrics import roc_auc_score

auc = roc_auc_score(y_test, y_pred_prob)
print("AUC:", auc)

In [64]:
y_pred_prob = model_lgbm.predict(X_test, num_iteration=model_lgbm.best_iteration)

#fpr, tpr, thresholds = roc_curve(y_test, y_pred_prob)
#j_scores = tpr - fpr
#best_index = np.argmax(j_scores)
#best_threshold = thresholds[best_index]
#print("Best Threshold: ", best_threshold)

from sklearn.metrics import f1_score

best_thresh = 0
best_accu = 0
t = 0.00
while t <= 1.00:
    t+=0.0001
    y_pred = (y_pred_prob > t).astype(int)
    accu = accuracy_score(y_test, y_pred)
    if accu > best_accu:
        best_accu = accu
        best_thresh = t

print(f"Best threshold (accu): {best_thresh}\nBest Accuracy: {best_accu}")

from sklearn.metrics import accuracy_score, classification_report, confusion_matrix, roc_auc_score
p = 0.4595
y_pred = (y_pred_prob > p).astype(int)

print(f"\nAccuracy: {accuracy_score(y_test, y_pred)}")
print("AUC:", roc_auc_score(y_test, y_pred_prob))
print("Classification Report:\n", classification_report(y_test, y_pred, target_names=['dogs','humans']))
print("Confusion Matrix:\n", confusion_matrix(y_test, y_pred))


Best threshold (accu): 0.4548999999999662
Best Accuracy: 0.7917102481361967

Accuracy: 0.7913319238900635
AUC: 0.8866891883251419
Classification Report:
               precision    recall  f1-score   support

        dogs       0.81      0.72      0.76     41701
      humans       0.78      0.86      0.81     48169

    accuracy                           0.79     89870
   macro avg       0.79      0.79      0.79     89870
weighted avg       0.79      0.79      0.79     89870

Confusion Matrix:
 [[29850 11851]
 [ 6902 41267]]


p = 0.4595
Accuracy: 0.7875486814287304

Best threshold (F1): 0.3785999999999746
Accuracy: 0.7913319238900635

Best threshold (F1): 0.45600000000000035
Best Accuracy: 0.7917102481361967
Accuracy: 0.7913319238900635
AUC: 0.8866891883251419

plt.plot(fpr, tpr, label="ROC Curve")
plt.xlabel("False Positive Rate")
plt.ylabel("True Positive Rate")
plt.title("ROC Curve")
plt.legend()
plt.show()

In [ ]:
y_pred = rf.predict(X_test)
breaknum_iteration=model_lgbm.best_iterationaccuracy = accuracy_score(y_test, y_pred)
print(f"\nAccuracy: {accuracy:.3f}")

species_names = ['Dog','Human']
print("\nClassification Report:")
print(classification_report(y_test, y_pred, target_names = species_names))

In [ ]:
num_iteration=model_lgbm.best_iteration

In [ ]:
print(df['species'].value_counts())
print(len(df))